# SI370 Day 2: (Re-) Introduction to Pandas


Notebook version 1.3

Note: The course git repo can be accessed via https://gitlab.si.umich.edu/jupyter/si370

# Reminders
* Slack (see link via Canvas announcements)
* survey

## Learning Objectives

* Pandas refresher (or introduction)
* explain how pandas operations differ from "traditional" python
* be able to load a CSV file into a Pandas DataFrame
* explain how to extract columns from a DataFrame
* sort a DataFrame
* assign a column as the index of a DataFrame
* filter a DataFrame according to some criteria
* explain how boolean masks work in filtering DataFrames

### IMPORTANT: Replace ```?``` in the following code with your uniqname.

In [1]:
MY_UNIQNAME = 'ktops'

## <font color="magenta">Exercise 1 (2 points): Consider Leek's Figure 2.1:</font>

![EDA](resources/leek2.1.png)

Do you agree with this layout of data analyses?  Can you think of ways to improve it?  Should Exploratory Data Analysis be constrained to where they are in the diagram?  Use a whiteboard to draw a better diagram.  Consider taking a picture of the whiteboard and inserting it in this notebook.

I think the layout doesn't allow for more than one way to get to different forms of analysis. I also don't think data analysis is this black and white in terms of one aspect of the work determining the entire type of analysis, but I could be wrong about that. I don't think these methods are mutually exclusive either, so I would recreate it in a way that doesn't haveit building off of itself like there's only one way to do something and you can only have done one type of analysis at a time. I don't think exploratory data analysis should be confined to where it is, you can be multiple things while being exploratory, ex: descriptive and exploratory.

From Chris: the whole thing here is exploratory analysis. These are more like steps in exploratory data than different things. Might make us remake this for the exam!

## Pandas
* high-level library to support data manipulation and analysis
* DataFrame is the primary object we’ll be dealing with
* similar to R’s dataframe
* maps onto tabular structure
* good for time series and econometric data

## Shift from "pythonic" to "pandorable"

* less looping over elements
* lots of built-in functionality
* a "paradigm shift"

# Data structures

We're all familiar with lists:

In [2]:
names = ["Charlotte", "Ingrid", "Ian", "Eric"]
scores = [80, 95, 85, 70]

Now let's say that we wanted to divide each of those scores by two and assign the results to another variable. Go ahead and write some code that does that... There are lots of ways to do this, so go ahead and write one way to do it (without importing any additional python packages) and assign the results to a 
variable called ```half```:

In [3]:
# import pandas as pd
# scores_df = pd.DataFrame(scores)
# half = scores_df.apply(lambda x: x/2)
# half

half = []
for score in scores:
    half.append(score/2)

If you followed the above instructions, the following cell block should print
a list of floats that looks like ```
[40.0, 47.5, 42.5, 35.0]```


In [4]:
half

[40.0, 47.5, 42.5, 35.0]

We can put data into an array structure that allows us to apply more powerful
functions.  The data structure that we're interested in is called an ```ndarray``` and is from the ```numpy``` package:

In [5]:
import numpy as np
ascores = np.array(scores)

In [6]:
ascores 

array([80, 95, 85, 70])

In [7]:
ahalf = ascores / 2

In [8]:
ahalf

array([40. , 47.5, 42.5, 35. ])

Numpy arrays are powerful, but they have some limitations:  they can only 
consist of one type of data (e.g. int), etc.  pandas provides two additional
data structures that are built on numpy ndarrays.

The first are Series.  Let's create a simple pandas Series and examine it:

In [9]:
import pandas as pd

In [10]:
from pandas import Series

In [11]:
sscores = Series(scores,name='scores')

In [12]:
sscores

0    80
1    95
2    85
3    70
Name: scores, dtype: int64

So you see a couple of useful things: an index (0 to 3) and a data type (dtype), which in this case is an int64.

**A Series is a one-dimensional ndarray with axis labels**

In [13]:
data = dict(zip(names,scores))

In [14]:
import pandas as pd

In [15]:
data

{'Charlotte': 80, 'Ingrid': 95, 'Ian': 85, 'Eric': 70}

In [16]:
sData = Series(data=data,name='score')

In [17]:
sData

Charlotte    80
Ingrid       95
Ian          85
Eric         70
Name: score, dtype: int64

So Series are a bit friendlier than numpy arrays, but they're still only one-dimensional.  Keep in mind that our basic data abstraction is a table, which can
be thought of as a two-dimensional array.  Let's go ahead and create a simple DataFrame with just one column:

In [18]:
from pandas import DataFrame


In [19]:
sData.to_frame()

score
Charlotte     80
Ingrid        95
Ian           85
Eric          70

Let's return to the code we ran last time and walk though it just to make sure we understand it

In [20]:
years = range(1880, 2015)
pieces = []
for year in years:
    path = 'data/names/yob%d.csv'%year
    frame = pd.read_csv(path)
    frame['year'] = year
    pieces.append(frame)
df_names = pd.concat(pieces, ignore_index=True)

## Today's focus: filtering, slicing and dicing

We're going to use some data from the [Nutrition Facts for McDonald's Menu](https://www.kaggle.com/mcdonalds/nutrition-facts) dataset on [Kaggle](www.kaggle.com).

Go ahead and browse the file using Jupyter.

Now let's load the file using ```read_csv```.

In [21]:
import pandas as pd

In [22]:
menu = pd.read_csv('data/menu.csv')

## <font color="magenta"> Exercise 2 (1 point): How many rows and columns are in this dataset?  Include one cell block to determine the number and one markdown block that presents the answer as a complete sentence (i.e. "The McDonald's nutrition data set contains X rows and Y columns"). </font>

In [23]:
numrows = len(menu)
numcols = len(menu.columns)
print(numrows)
print(numcols)

260
24


The McDonald's nutrition data set contains 260 rows and 24 columns.


# Headings
## Subheading

* one
* two 
* three


1. first
1. other
2. blah
2. foo


## Extracting columns 

Getting column names is easy:

In [24]:
menu.columns

Index(['Category', 'Item', 'Serving Size', 'Calories', 'Calories from Fat',
       'Total Fat', 'Total Fat (% Daily Value)', 'Saturated Fat',
       'Saturated Fat (% Daily Value)', 'Trans Fat', 'Cholesterol',
       'Cholesterol (% Daily Value)', 'Sodium', 'Sodium (% Daily Value)',
       'Carbohydrates', 'Carbohydrates (% Daily Value)', 'Dietary Fiber',
       'Dietary Fiber (% Daily Value)', 'Sugars', 'Protein',
       'Vitamin A (% Daily Value)', 'Vitamin C (% Daily Value)',
       'Calcium (% Daily Value)', 'Iron (% Daily Value)'],
      dtype='object')

Similarly, extracting a specific columns is also easy:

In [25]:
menu['Category'] 
menu.Category

0               Breakfast
1               Breakfast
2               Breakfast
3               Breakfast
4               Breakfast
5               Breakfast
6               Breakfast
7               Breakfast
8               Breakfast
9               Breakfast
10              Breakfast
11              Breakfast
12              Breakfast
13              Breakfast
14              Breakfast
15              Breakfast
16              Breakfast
17              Breakfast
18              Breakfast
19              Breakfast
20              Breakfast
21              Breakfast
22              Breakfast
23              Breakfast
24              Breakfast
25              Breakfast
26              Breakfast
27              Breakfast
28              Breakfast
29              Breakfast
              ...        
230          Coffee & Tea
231          Coffee & Tea
232    Smoothies & Shakes
233    Smoothies & Shakes
234    Smoothies & Shakes
235    Smoothies & Shakes
236    Smoothies & Shakes
237    Smoot

And multiple columns can also be extracted by passing a list of column names

In [26]:
menu[['Item','Calories']]

Item  Calories
0                                         Egg McMuffin       300
1                                    Egg White Delight       250
2                                     Sausage McMuffin       370
3                            Sausage McMuffin with Egg       450
4                     Sausage McMuffin with Egg Whites       400
5                                 Steak & Egg McMuffin       430
6        Bacon, Egg & Cheese Biscuit (Regular Biscuit)       460
7          Bacon, Egg & Cheese Biscuit (Large Biscuit)       520
8    Bacon, Egg & Cheese Biscuit with Egg Whites (R...       410
9    Bacon, Egg & Cheese Biscuit with Egg Whites (L...       470
10                   Sausage Biscuit (Regular Biscuit)       430
11                     Sausage Biscuit (Large Biscuit)       480
12          Sausage Biscuit with Egg (Regular Biscuit)       510
13            Sausage Biscuit with Egg (Large Biscuit)       570
14   Sausage Biscuit with Egg Whites (Regular Biscuit)       460
15     Sausage Biscuit with Egg Whites (Large Biscuit)       520
16    Southern Style Chicken Biscuit (Regular Biscuit)       410
17      Southern Style Chicken Biscuit (Large Biscuit)       470
18               Steak & Egg Biscuit (Regular Biscuit)       540
19                      Bacon, Egg & Cheese McGriddles       460
20      Bacon, Egg & Cheese McGriddles with Egg Whites       400
21                                  Sausage McGriddles       420
22                    Sausage, Egg & Cheese McGriddles       550
23    Sausage, Egg & Cheese McGriddles with Egg Whites       500
24                           Bacon, Egg & Cheese Bagel       620
25           Bacon, Egg & Cheese Bagel with Egg Whites       570
26                           Steak, Egg & Cheese Bagel       670
27                     Big Breakfast (Regular Biscuit)       740
28                       Big Breakfast (Large Biscuit)       800
29     Big Breakfast with Egg Whites (Regular Biscuit)       640
..                                                 ...       ...
230                     Frappé Chocolate Chip (Medium)       630
231                      Frappé Chocolate Chip (Large)       760
232             Blueberry Pomegranate Smoothie (Small)       220
233            Blueberry Pomegranate Smoothie (Medium)       260
234             Blueberry Pomegranate Smoothie (Large)       340
235                 Strawberry Banana Smoothie (Small)       210
236                Strawberry Banana Smoothie (Medium)       250
237                 Strawberry Banana Smoothie (Large)       330
238                   Mango Pineapple Smoothie (Small)       210
239                  Mango Pineapple Smoothie (Medium)       260
240                   Mango Pineapple Smoothie (Large)       340
241                              Vanilla Shake (Small)       530
242                             Vanilla Shake (Medium)       660
243                              Vanilla Shake (Large)       820
244                           Strawberry Shake (Small)       550
245                          Strawberry Shake (Medium)       690
246                           Strawberry Shake (Large)       850
247                            Chocolate Shake (Small)       560
248                           Chocolate Shake (Medium)       700
249                            Chocolate Shake (Large)       850
250                            Shamrock Shake (Medium)       660
251                             Shamrock Shake (Large)       820
252                McFlurry with M&M’s Candies (Small)       650
253               McFlurry with M&M’s Candies (Medium)       930
254                McFlurry with M&M’s Candies (Snack)       430
255                 McFlurry with Oreo Cookies (Small)       510
256                McFlurry with Oreo Cookies (Medium)       690
257                 McFlurry with Oreo Cookies (Snack)       340
258  McFlurry with Reese's Peanut Butter Cups (Medium)       810
259   McFlurry with Reese's Peanut Butter Cups (Snack)       410

[260 rows x 2 colum

## Extracting rows

In [27]:
menu.iloc[0]

Category                              Breakfast
Item                               Egg McMuffin
Serving Size                     4.8 oz (136 g)
Calories                                    300
Calories from Fat                           120
Total Fat                                    13
Total Fat (% Daily Value)                    20
Saturated Fat                                 5
Saturated Fat (% Daily Value)                25
Trans Fat                                     0
Cholesterol                                 260
Cholesterol (% Daily Value)                  87
Sodium                                      750
Sodium (% Daily Value)                       31
Carbohydrates                                31
Carbohydrates (% Daily Value)                10
Dietary Fiber                                 4
Dietary Fiber (% Daily Value)                17
Sugars                                        3
Protein                                      17
Vitamin A (% Daily Value)               

In [28]:
menu

Category                                               Item  \
0             Breakfast                                       Egg McMuffin   
1             Breakfast                                  Egg White Delight   
2             Breakfast                                   Sausage McMuffin   
3             Breakfast                          Sausage McMuffin with Egg   
4             Breakfast                   Sausage McMuffin with Egg Whites   
5             Breakfast                               Steak & Egg McMuffin   
6             Breakfast      Bacon, Egg & Cheese Biscuit (Regular Biscuit)   
7             Breakfast        Bacon, Egg & Cheese Biscuit (Large Biscuit)   
8             Breakfast  Bacon, Egg & Cheese Biscuit with Egg Whites (R...   
9             Breakfast  Bacon, Egg & Cheese Biscuit with Egg Whites (L...   
10            Breakfast                  Sausage Biscuit (Regular Biscuit)   
11            Breakfast                    Sausage Biscuit (Large Biscuit)   
12            Breakfast         Sausage Biscuit with Egg (Regular Biscuit)   
13            Breakfast           Sausage Biscuit with Egg (Large Biscuit)   
14            Breakfast  Sausage Biscuit with Egg Whites (Regular Biscuit)   
15            Breakfast    Sausage Biscuit with Egg Whites (Large Biscuit)   
16            Breakfast   Southern Style Chicken Biscuit (Regular Biscuit)   
17            Breakfast     Southern Style Chicken Biscuit (Large Biscuit)   
18            Breakfast              Steak & Egg Biscuit (Regular Biscuit)   
19            Breakfast                     Bacon, Egg & Cheese McGriddles   
20            Breakfast     Bacon, Egg & Cheese McGriddles with Egg Whites   
21            Breakfast                                 Sausage McGriddles   
22            Breakfast                   Sausage, Egg & Cheese McGriddles   
23            Breakfast   Sausage, Egg & Cheese McGriddles with Egg Whites   
24            Breakfast                          Bacon, Egg & Cheese Bagel   
25            Breakfast          Bacon, Egg & Cheese Bagel with Egg Whites   
26            Breakfast                          Steak, Egg & Cheese Bagel   
27            Breakfast                    Big Breakfast (Regular Biscuit)   
28            Breakfast                      Big Breakfast (Large Biscuit)   
29            Breakfast    Big Breakfast with Egg Whites (Regular Biscuit)   
..                  ...                                                ...   
230        Coffee & Tea                     Frappé Chocolate Chip (Medium)   
231        Coffee & Tea                      Frappé Chocolate Chip (Large)   
232  Smoothies & Shakes             Blueberry Pomegranate Smoothie (Small)   
233  Smoothies & Shakes            Blueberry Pomegranate Smoothie (Medium)   
234  Smoothies & Shakes             Blueberry Pomegranate Smoothie (Large)   
235  Smoothies & Shakes                 Strawberry Banana Smoothie (Small)   
236  Smoothies & Shakes                Strawberry Banana Smoothie (Medium)   
237  Smoothies & Shakes                 Strawberry Banana Smoothie (Large)   
238  Smoothies & Shakes                   Mango Pineapple Smoothie (Small)   
239  Smoothies & Shakes                  Mango Pineapple Smoothie (Medium)   
240  Smoothies & Shakes                   Mango Pineapple Smoothie (Large)   
241  Smoothies & Shakes                              Vanilla Shake (Small)   
242  Smoothies & Shakes                             Vanilla Shake (Medium)   
243  Smoothies & Shakes                              Vanilla Shake (Large)   
244  Smoothies & Shakes                           Strawberry Shake (Small)   
245  Smoothies & Shakes                          Strawberry Shake (Medium)   
246  Smoothies & Shakes                           Strawberry Shake (Large)   
247  Smoothies & Shakes                            Chocolate Shake (Small)   
248  Smoothies & Shakes                           Chocolate Shake (Medium)   
249  Smoothies & Shakes              

You'll notice that the index column is just a series of integers starting with 0.  Sometimes that's fine.  
Other times we want to assign a more useful row as the index.  Note that the values in the index do not need to be unique.

In [29]:
menu_i = menu.set_index('Item')

In [30]:
menu_i

Category  \
Item                                                                     
Egg McMuffin                                                 Breakfast   
Egg White Delight                                            Breakfast   
Sausage McMuffin                                             Breakfast   
Sausage McMuffin with Egg                                    Breakfast   
Sausage McMuffin with Egg Whites                             Breakfast   
Steak & Egg McMuffin                                         Breakfast   
Bacon, Egg & Cheese Biscuit (Regular Biscuit)                Breakfast   
Bacon, Egg & Cheese Biscuit (Large Biscuit)                  Breakfast   
Bacon, Egg & Cheese Biscuit with Egg Whites (Re...           Breakfast   
Bacon, Egg & Cheese Biscuit with Egg Whites (La...           Breakfast   
Sausage Biscuit (Regular Biscuit)                            Breakfast   
Sausage Biscuit (Large Biscuit)                              Breakfast   
Sausage Biscuit with Egg (Regular Biscuit)                   Breakfast   
Sausage Biscuit with Egg (Large Biscuit)                     Breakfast   
Sausage Biscuit with Egg Whites (Regular Biscuit)            Breakfast   
Sausage Biscuit with Egg Whites (Large Biscuit)              Breakfast   
Southern Style Chicken Biscuit (Regular Biscuit)             Breakfast   
Southern Style Chicken Biscuit (Large Biscuit)               Breakfast   
Steak & Egg Biscuit (Regular Biscuit)                        Breakfast   
Bacon, Egg & Cheese McGriddles                               Breakfast   
Bacon, Egg & Cheese McGriddles with Egg Whites               Breakfast   
Sausage McGriddles                                           Breakfast   
Sausage, Egg & Cheese McGriddles                             Breakfast   
Sausage, Egg & Cheese McGriddles with Egg Whites             Breakfast   
Bacon, Egg & Cheese Bagel                                    Breakfast   
Bacon, Egg & Cheese Bagel with Egg Whites                    Breakfast   
Steak, Egg & Cheese Bagel                                    Breakfast   
Big Breakfast (Regular Biscuit)                              Breakfast   
Big Breakfast (Large Biscuit)                                Breakfast   
Big Breakfast with Egg Whites (Regular Biscuit)              Breakfast   
...                                                                ...   
Frappé Chocolate Chip (Medium)                            Coffee & Tea   
Frappé Chocolate Chip (Large)                             Coffee & Tea   
Blueberry Pomegranate Smoothie (Small)              Smoothies & Shakes   
Blueberry Pomegranate Smoothie (Medium)             Smoothies & Shakes   
Blueberry Pomegranate Smoothie (Large)              Smoothies & Shakes   
Strawberry Banana Smoothie (Small)                  Smoothies & Shakes   
Strawberry Banana Smoothie (Medium)                 Smoothies & Shakes   
Strawberry Banana Smoothie (Large)                  Smoothies & Shakes   
Mango Pineapple Smoothie (Small)                    Smoothies & Shakes   
Mango Pineapple Smoothie (Medium)                   Smoothies & Shakes   
Mango Pineapple Smoothie (Large)                    Smoothies & Shakes   
Vanilla Shake (Small)                               Smoothies & Shakes   
Vanilla Shake (Medium)                              Smoothies & Shakes   
Vanilla Shake (Large)                               Smoothies & Shakes   
Strawberry Shake (Small)                            Smoothies & Shakes   
Strawberry Shake (Medium)                           Smoothies & Shakes   
Strawberry Shake (Large)                            Smoothies & Shakes   
Chocolate Shake (Small)                             Smoothies & Shakes   
Chocolate Shake (Medium)                            Smoothies & Shakes   
Chocolate Shake (Large)                             Smoothies & Shakes   
Shamrock Shake (Medium)                             Smoothies & Shakes   
Shamrock Shake (Large)                              Smoothies & Sh

In [31]:
menu_i.loc[0]

TypeError: cannot do label indexing on <class 'pandas.core.indexes.base.Index'> with these indexers [0] of <class 'int'>

In [32]:
menu_i.iloc[0]

Category                              Breakfast
Serving Size                     4.8 oz (136 g)
Calories                                    300
Calories from Fat                           120
Total Fat                                    13
Total Fat (% Daily Value)                    20
Saturated Fat                                 5
Saturated Fat (% Daily Value)                25
Trans Fat                                     0
Cholesterol                                 260
Cholesterol (% Daily Value)                  87
Sodium                                      750
Sodium (% Daily Value)                       31
Carbohydrates                                31
Carbohydrates (% Daily Value)                10
Dietary Fiber                                 4
Dietary Fiber (% Daily Value)                17
Sugars                                        3
Protein                                      17
Vitamin A (% Daily Value)                    10
Vitamin C (% Daily Value)               

We can also extract a row and a slice of its columns

In [33]:
menu_i.iloc[0,0:2]

Category             Breakfast
Serving Size    4.8 oz (136 g)
Name: Egg McMuffin, dtype: object

Or we can extract a column and a slice of its rows


In [34]:
menu_i.iloc[1:3,:]

Category    Serving Size  Calories  Calories from Fat  \
Item                                                                        
Egg White Delight  Breakfast  4.8 oz (135 g)       250                 70   
Sausage McMuffin   Breakfast  3.9 oz (111 g)       370                200   

                   Total Fat  Total Fat (% Daily Value)  Saturated Fat  \
Item                                                                     
Egg White Delight        8.0                         12            3.0   
Sausage McMuffin        23.0                         35            8.0   

                   Saturated Fat (% Daily Value)  Trans Fat  Cholesterol  ...  \
Item                                                                      ...   
Egg White Delight                             15        0.0           25  ...   
Sausage McMuffin                              42        0.0           45  ...   

                   Carbohydrates  Carbohydrates (% Daily Value)  \
Item                                                              
Egg White Delight             30                             10   
Sausage McMuffin              29                             10   

                   Dietary Fiber  Dietary Fiber (% Daily Value)  Sugars  \
Item                                                                      
Egg White Delight              4                             17       3   
Sausage McMuffin               4                             17       2   

                   Protein  Vitamin A (% Daily Value)  \
Item                                                    
Egg White Delight       18                          6   
Sausage McMuffin        14                          8   

                   Vitamin C (% Daily Value)  Calcium (% Daily Value)  \
Item                                                                    
Egg White Delight                          0                       25   
Sausage McMuffin                           0                       25   

                   Iron (% Daily Value)  
Item                                     
Egg White Delight                     8  
Sausage McMuffin                     10  

[2 rows x 23 columns]

## Sorting
Sorting is supported using sort_index and sort_values:


In [35]:
menu_sorted_by_cals = menu.sort_values('Calories',ascending=True)

## <font color="magenta"> Exercise 3 (2 points): Display the four menu items that have the most Saturated Fat (the absolute amount, not the % Daily Value).</font>

In [36]:
menu_sorted_by_fat = menu.sort_values('Saturated Fat', ascending=False)
menu_sorted_by_fat.head(4)

Category                                         Item  \
231        Coffee & Tea                Frappé Chocolate Chip (Large)   
82       Chicken & Fish                 Chicken McNuggets (40 piece)   
32            Breakfast  Big Breakfast with Hotcakes (Large Biscuit)   
253  Smoothies & Shakes         McFlurry with M&M’s Candies (Medium)   

        Serving Size  Calories  Calories from Fat  Total Fat  \
231     22 fl oz cup       760                280       31.0   
82   22.8 oz (646 g)      1880               1060      118.0   
32   15.3 oz (434 g)      1150                540       60.0   
253  16.2 oz (460 g)       930                290       33.0   

     Total Fat (% Daily Value)  Saturated Fat  Saturated Fat (% Daily Value)  \
231                         48           20.0                            101   
82                         182           20.0                            101   
32                          93           20.0                            100   
253                         50           20.0                            102   

     Trans Fat  ...  Carbohydrates  Carbohydrates (% Daily Value)  \
231        1.5  ...            111                             37   
82         1.0  ...            118                             39   
32         0.0  ...            116                             39   
253        1.0  ...            139                             46   

     Dietary Fiber  Dietary Fiber (% Daily Value)  Sugars  Protein  \
231              1                              5      99       12   
82               6                             24       1       87   
32               7                             28      17       36   
253              2                              7     128       20   

     Vitamin A (% Daily Value)  Vitamin C (% Daily Value)  \
231                         20                          0   
82                           0                         15   
32                          15                          2   
253                         25                          0   

     Calcium (% Daily Value)  Iron (% Daily Value)  
231                       35                     6  
82                         8                    25  
32                        30                    40  
253                       70                    10  

[4 rows x 24 columns]

## Filtering

More often than extracting a particular row, we want to extract one or more rows that match
some criteria.  For example, to find all the menu items that contain Trans Fats, we could use:


In [37]:
menu['Trans Fat' ] > 0

0      False
1      False
2      False
3      False
4      False
5       True
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18      True
19     False
20     False
21     False
22     False
23     False
24      True
25      True
26      True
27     False
28     False
29     False
       ...  
230     True
231     True
232    False
233    False
234    False
235    False
236    False
237    False
238    False
239    False
240    False
241     True
242     True
243     True
244     True
245     True
246     True
247     True
248     True
249     True
250     True
251     True
252     True
253     True
254    False
255     True
256     True
257    False
258     True
259    False
Name: Trans Fat, Length: 260, dtype: bool

In [38]:
menu_trans_fats = menu[menu['Trans Fat'] > 0.0]

We're going to spend time in class explaining what just happened there.

In [39]:
menu['Trans Fat']

0      0.0
1      0.0
2      0.0
3      0.0
4      0.0
5      1.0
6      0.0
7      0.0
8      0.0
9      0.0
10     0.0
11     0.0
12     0.0
13     0.0
14     0.0
15     0.0
16     0.0
17     0.0
18     1.0
19     0.0
20     0.0
21     0.0
22     0.0
23     0.0
24     0.5
25     0.5
26     1.5
27     0.0
28     0.0
29     0.0
      ... 
230    1.0
231    1.5
232    0.0
233    0.0
234    0.0
235    0.0
236    0.0
237    0.0
238    0.0
239    0.0
240    0.0
241    1.0
242    1.0
243    1.0
244    1.0
245    1.0
246    1.0
247    1.0
248    1.0
249    1.0
250    1.0
251    1.0
252    0.5
253    1.0
254    0.0
255    0.5
256    1.0
257    0.0
258    1.0
259    0.0
Name: Trans Fat, Length: 260, dtype: float64

In [40]:
menu['Trans Fat'] > 0.0

0      False
1      False
2      False
3      False
4      False
5       True
6      False
7      False
8      False
9      False
10     False
11     False
12     False
13     False
14     False
15     False
16     False
17     False
18      True
19     False
20     False
21     False
22     False
23     False
24      True
25      True
26      True
27     False
28     False
29     False
       ...  
230     True
231     True
232    False
233    False
234    False
235    False
236    False
237    False
238    False
239    False
240    False
241     True
242     True
243     True
244     True
245     True
246     True
247     True
248     True
249     True
250     True
251     True
252     True
253     True
254    False
255     True
256     True
257    False
258     True
259    False
Name: Trans Fat, Length: 260, dtype: bool

In [41]:
menu[menu['Trans Fat'] > 0.0]

Category                                               Item  \
5             Breakfast                               Steak & Egg McMuffin   
18            Breakfast              Steak & Egg Biscuit (Regular Biscuit)   
24            Breakfast                          Bacon, Egg & Cheese Bagel   
25            Breakfast          Bacon, Egg & Cheese Bagel with Egg Whites   
26            Breakfast                          Steak, Egg & Cheese Bagel   
42          Beef & Pork                                            Big Mac   
43          Beef & Pork                        Quarter Pounder with Cheese   
44          Beef & Pork                Quarter Pounder with Bacon & Cheese   
45          Beef & Pork          Quarter Pounder with Bacon Habanero Ranch   
46          Beef & Pork                             Quarter Pounder Deluxe   
47          Beef & Pork                 Double Quarter Pounder with Cheese   
49          Beef & Pork                                       Cheeseburger   
50          Beef & Pork                                Double Cheeseburger   
51          Beef & Pork                             Bacon Clubhouse Burger   
52          Beef & Pork                                           McDouble   
53          Beef & Pork                                     Bacon McDouble   
54          Beef & Pork                                       Daily Double   
55          Beef & Pork                                    Jalapeño Double   
63       Chicken & Fish            Bacon Clubhouse Crispy Chicken Sandwich   
70       Chicken & Fish    Premium McWrap Chicken & Bacon (Crispy Chicken)   
72       Chicken & Fish    Premium McWrap Chicken & Ranch (Crispy Chicken)   
73       Chicken & Fish   Premium McWrap Chicken & Ranch (Grilled Chicken)   
74       Chicken & Fish  Premium McWrap Southwest Chicken (Crispy Chicken)   
82       Chicken & Fish                       Chicken McNuggets (40 piece)   
180        Coffee & Tea                                      Mocha (Large)   
185        Coffee & Tea                             Caramel Mocha (Medium)   
186        Coffee & Tea                              Caramel Mocha (Large)   
191        Coffee & Tea                             Hot Chocolate (Medium)   
192        Coffee & Tea                              Hot Chocolate (Large)   
213        Coffee & Tea                                 Iced Mocha (Large)   
219        Coffee & Tea                         Iced Caramel Mocha (Large)   
223        Coffee & Tea                               Frappé Mocha (Small)   
224        Coffee & Tea                              Frappé Mocha (Medium)   
225        Coffee & Tea                               Frappé Mocha (Large)   
226        Coffee & Tea                             Frappé Caramel (Small)   
227        Coffee & Tea                            Frappé Caramel (Medium)   
228        Coffee & Tea                             Frappé Caramel (Large)   
229        Coffee & Tea                      Frappé Chocolate Chip (Small)   
230        Coffee & Tea                     Frappé Chocolate Chip (Medium)   
231        Coffee & Tea                      Frappé Chocolate Chip (Large)   
241  Smoothies & Shakes                              Vanilla Shake (Small)   
242  Smoothies & Shakes                             Vanilla Shake (Medium)   
243  Smoothies & Shakes                              Vanilla Shake (Large)   
244  Smoothies & Shakes                           Strawberry Shake (Small)   
245  Smoothies & Shakes                          Strawberry Shake (Medium)   
246  Smoothies & Shakes                           Strawberry Shake (Large)   
247  Smoothies & Shakes                            Chocolate Shake (Small)   
248  Smoothies & Shakes                           Chocolate Shake (Medium)   
249  Smoothies & Shakes                            Chocolate Shake (Large)   
250  Smoothies & Shakes                            Shamrock Shake (Medium)   
251  Smoothies & Shakes              

In [42]:
menu.columns

Index(['Category', 'Item', 'Serving Size', 'Calories', 'Calories from Fat',
       'Total Fat', 'Total Fat (% Daily Value)', 'Saturated Fat',
       'Saturated Fat (% Daily Value)', 'Trans Fat', 'Cholesterol',
       'Cholesterol (% Daily Value)', 'Sodium', 'Sodium (% Daily Value)',
       'Carbohydrates', 'Carbohydrates (% Daily Value)', 'Dietary Fiber',
       'Dietary Fiber (% Daily Value)', 'Sugars', 'Protein',
       'Vitamin A (% Daily Value)', 'Vitamin C (% Daily Value)',
       'Calcium (% Daily Value)', 'Iron (% Daily Value)'],
      dtype='object')

## <font color="magenta">Exercise 4 (2 points): List the top 3 breakfast items have the most Dietary Fiber.</font>

In [43]:
fiber_menu = menu[menu['Category'] == 'Breakfast'].sort_values('Dietary Fiber', ascending = False)
fiber_menu.head(3)

Category                                               Item  \
34  Breakfast  Big Breakfast with Hotcakes and Egg Whites (La...   
32  Breakfast        Big Breakfast with Hotcakes (Large Biscuit)   
33  Breakfast  Big Breakfast with Hotcakes and Egg Whites (Re...   

       Serving Size  Calories  Calories from Fat  Total Fat  \
34  15.4 oz (437 g)      1050                450       50.0   
32  15.3 oz (434 g)      1150                540       60.0   
33  14.9 oz (423 g)       990                410       46.0   

    Total Fat (% Daily Value)  Saturated Fat  Saturated Fat (% Daily Value)  \
34                         77           16.0                             81   
32                         93           20.0                            100   
33                         70           16.0                             78   

    Trans Fat  ...  Carbohydrates  Carbohydrates (% Daily Value)  \
34        0.0  ...            115                             38   
32        0.0  ...            116                             39   
33        0.0  ...            110                             37   

    Dietary Fiber  Dietary Fiber (% Daily Value)  Sugars  Protein  \
34              7                             28      18       35   
32              7                             28      17       36   
33              6                             23      17       35   

    Vitamin A (% Daily Value)  Vitamin C (% Daily Value)  \
34                          4                          2   
32                         15                          2   
33                          0                          2   

    Calcium (% Daily Value)  Iron (% Daily Value)  
34                       25                    30  
32                       30                    40  
33                       25                    30  

[3 rows x 24 columns]

## <font color="magenta">Exercise 5 (3 points): Show up to three of the best choices for someone who is following the "Atkin's Diet" (Google it).  Justify your choices in a markdown block that follows your code.</font>

In [44]:
atkins = menu.sort_values('Carbohydrates', ascending=True)
atkins.head(3)

Category                 Item  Serving Size  Calories  \
136     Beverages  Dasani Water Bottle    16.9 fl oz         0   
145  Coffee & Tea       Coffee (Small)  12 fl oz cup         0   
140  Coffee & Tea     Iced Tea (Child)  12 fl oz cup         0   

     Calories from Fat  Total Fat  Total Fat (% Daily Value)  Saturated Fat  \
136                  0        0.0                          0            0.0   
145                  0        0.0                          0            0.0   
140                  0        0.0                          0            0.0   

     Saturated Fat (% Daily Value)  Trans Fat  ...  Carbohydrates  \
136                              0        0.0  ...              0   
145                              0        0.0  ...              0   
140                              0        0.0  ...              0   

     Carbohydrates (% Daily Value)  Dietary Fiber  \
136                              0              0   
145                              0              0   
140                              0              0   

     Dietary Fiber (% Daily Value)  Sugars  Protein  \
136                              0       0        0   
145                              0       0        0   
140                              0       0        0   

     Vitamin A (% Daily Value)  Vitamin C (% Daily Value)  \
136                          0                          0   
145                          0                          0   
140                          0                          0   

     Calcium (% Daily Value)  Iron (% Daily Value)  
136                        0                     0  
145                        0                     0  
140                        0                     0  

[3 rows x 24 columns]

These are all drinks, let's get rid of 'Beverages' and 'Coffee & Tea'

In [48]:
atkins = menu.sort_values('Carbohydrates', ascending=True)
atkins = atkins.set_index('Category')
atkins = atkins.drop('Beverages')
atkins = atkins.drop('Coffee & Tea')
atkins.head(3)

Item   Serving Size  Calories  \
Category                                                       
Snacks & Sides         Apple Slices  1.2 oz (34 g)        15   
Snacks & Sides           Side Salad  3.1 oz (87 g)        20   
Desserts        Kids Ice Cream Cone    1 oz (29 g)        45   

                Calories from Fat  Total Fat  Total Fat (% Daily Value)  \
Category                                                                  
Snacks & Sides                  0        0.0                          0   
Snacks & Sides                  0        0.0                          0   
Desserts                       10        1.5                          2   

                Saturated Fat  Saturated Fat (% Daily Value)  Trans Fat  \
Category                                                                  
Snacks & Sides            0.0                              0        0.0   
Snacks & Sides            0.0                              0        0.0   
Desserts                  1.0                              4        0.0   

                Cholesterol  ...  Carbohydrates  \
Category                     ...                  
Snacks & Sides            0  ...              4   
Snacks & Sides            0  ...              4   
Desserts                  5  ...              7   

                Carbohydrates (% Daily Value)  Dietary Fiber  \
Category                                                       
Snacks & Sides                              1              0   
Snacks & Sides                              1              1   
Desserts                                    2              0   

                Dietary Fiber (% Daily Value)  Sugars  Protein  \
Category                                                         
Snacks & Sides                              0       3        0   
Snacks & Sides                              6       2        1   
Desserts                                    0       6        1   

                Vitamin A (% Daily Value)  Vitamin C (% Daily Value)  \
Category                                                               
Snacks & Sides                          0                        160   
Snacks & Sides                         45                         25   
Desserts                                2                          0   

                Calcium (% Daily Value)  Iron (% Daily Value)  
Category                                                       
Snacks & Sides                        2                     0  
Snacks & Sides                        2                     4  
Desserts                              4                     0  

[3 rows x 23 columns]

According to these menu choices, the best choices for someone on the Atkins diet (a no-carb diet) would be apple slices, a side salad, and a kids ice cream cone. I got these by ordering the options by least to most carbs, and then taking out the drink options so that we could see what food there was.

# <font color="magenta">END OF NOTEBOOK</font>
## Remember to submit this file in HTML format via Canvas.

iloc = integer location of a row